src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs

```diff
@@ -1,19 +1,336 @@
+
+            while (true)
+            {
+                if (reader == end)
+                {
+                    break;
+                }
+
+                if (data[reader] == '%')
+                {
+                    var decodeReader = reader;
+
+                    // If decoding process succeeds, the writer iterator will be moved
+                    // to the next write-ready location. On the other hand if the scanned
+                    // percent-encodings cannot be interpreted as sequence of UTF-8 octets,
+                    // these bytes should be copied to output as is. 
+                    // The decodeReader iterator is always moved to the first byte not yet 
+                    // be scanned after the process. A failed decoding means the chars
+                    // between the reader and decodeReader can be copied to output untouched. 
+                    if (!DecodeCore(ref decodeReader, ref writer, end, data))
+                    {
+                        Copy(reader, decodeReader, ref writer, data);
+                    }
+
+                    reader = decodeReader;
+                }
+                else
+                {
+                    data[writer++] = data[reader++];
+                }
+            }
+
+            return writer;
+        }
+
+        /// <summary>
+        /// Unescape a URL path
+        /// </summary>
+        /// <param name="source">The byte span represents a UTF8 encoding url path.</param>
+        /// <param name="destination">The byte span where unescaped url path is copied to.</param>
+        /// <returns>The length of the byte sequence of the unescaped url path.</returns>
+        public static int Unescape(Span<byte> source, Span<byte> destination)
+        {
+            if (destination.Length < source.Length)
+            {
+                throw new ArgumentException(
+                    "Lenghth of the destination byte span is less then the source.",
+                    nameof(destination));
+            }
+
+            // This requires the destination span to be larger or equal to source span
+            source.CopyTo(destination);
+            return UnescapeInPlace(destination);
+        }
+
+        /// <summary>
+        /// Unescape the percent-encodings
+        /// </summary>
+        /// <param name="reader">The iterator point to the first % char</param>
+        /// <param name="writer">The place to write to</param>
+        /// <param name="end">The end of the buffer</param>
+        /// <param name="buffer">The byte array</param>
+        private static bool DecodeCore(ref int reader, ref int writer, int end, Span<byte> buffer)
         {
+            // preserves the original head. if the percent-encodings cannot be interpreted as sequence of UTF-8 octets,
+            // bytes from this till the last scanned one will be copied to the memory pointed by writer.
+            var byte1 = UnescapePercentEncoding(ref reader, end, buffer);
```

Comment:
so this returns nullable byte. Do we have to do it? Can the calling code check the length of the buffer before calling this ? Would it be faster? I am asking because we need to be super careful with all these routines operating on single bytes. They can become quite expensive.